### ISSUE TITLE
Weird interaction between diff-hl-flydiff-mode and org-indent-mode

### ISSUE THREAD

**Issue by peniblec** [dgutov/diff-hl#142]
Hello,

I'm seeing a strange interaction between `diff-hl-flydiff-mode` and `org-indent-mode`.  When typing in the body of an Org section that is part of an added hunk, on every input, the line before point moves to the left margin.  It remains there as long as I keep typing; when I stop, the line moves back to `(* org-indent-indentation-per-level (org-current-level))` after a short delay.

To reproduce:

``` shell
#!/bin/bash

git init repro
cd repro
echo "* heading" > repro.org
git add repro.org
git commit -m "repro"
emacs -Q repro.org
```

Then:

- `M-x package-initialize`
- `M-x diff-hl-mode`
- `M-x diff-hl-flydiff-mode`
- `M-x org-indent-mode`
- `M->`
- `a RET a`

Inputting the second `a` causes the first `a` to be shoved to the left margin.

I have no idea whether things ought to be fixed in diff-hl or in org-indent; I could not figure out what causes this.  I don't mind bringing this to the Org mailing list once I understand what's happening…

Reproduced with

- emacs 26.3 and 28
- diff-hl 20200510.1540

Let me know if there is anything I can do to help debug this.

**Comment by dgutov** [comment]
Thank you. I can see the problem, but I don't really know how `org-indent` works.

Investigation is also complicated by the fact that I can't repro this in my personal configuration: it simply doesn't jump (with flydiff and org-indent enabled, of course).

The problem is likely in the interplay of these two features, and maybe there's no particular "fault" in either of them. I'd have to investigate deeper to say more.

**Comment by peniblec** [comment]
FWIW[^1], removing `diff-hl-flydiff-mode`'s `:override #'ignore` advice on `diff-hl-overlay-modified` makes the jumping stop; what happens instead is that diff-hl's fringe highlighting disappears while I type, and re-appears after a small delay once I stop typing.

[^1]: which is probably not much, since I'm pretty far into "commenting out random stuff hoping things will start making sense" territory; I just figured I'd drop this factlet before hitting the hay.

**Comment by dgutov** [comment]
Yup. This brings back `diff-hl`'s basic behavior: deleting "outdated" overlays.

Not sure how it relates to the problem, though.

**Comment by peniblec** [comment]
I've tried to `M-x debug-on-entry RET self-insert-command`; I've stepped through this function, `after-change-functions` and `post-self-insert-hook` as carefully as I could, but the bug does not show up while debugging.

I've also tried `(setq-local after-change-functions nil)`; this causes the *current line* to jitter, instead of the previous one.  That made me want to try to inspect these functions; since stepping through does not show anything, I've advised each function as follows, in order to see when the previous line gets shoved to the left:

``` elisp
(defun my/dbg (prefix f)
  (lambda (&rest r)
    (message "%s %s" prefix (symbol-name f))
    (redisplay)
    (sleep-for 1.0)))

(dolist (f after-change-functions)
  (when (functionp f)
    (advice-add f :before (my/dbg "before" f))
    (advice-add f :after (my/dbg "after" f))))
```

AFAICT, there are two functions that cause the previous line to be shifted left (meaning as long as one of them is in `after-change-functions`, the shift happens on the previous line).  I've tried to isolate which statements cause the shift and I think it comes down to:

- in `jit-lock-after-change`, the call to `put-text-property`,
- in `org-indent-refresh-maybe`, down the call stack in `org-indent-set-line-properties`, the call to `add-text-properties`.

Before those statements, the current line is shifted left; after whichever runs first, the previous line is shifted left.

If `diff-hl-flydiff-mode` is disabled, no shift happens.

Digging down `add_text_properties_1`, I see that `signal_after_change` eventually calls `report_overlay_modification`, which I'm going to assume eventually calls the hooks set on diff-hl's overlays, which are set to `diff-hl-overlay-modified`, which `diff-hl-flydiff-mode` overrides to `ignore`, and… I'm stumped.

Maybe I should `M-x report-emacs-bug`?

**Comment by dgutov** [comment]
Yeah, maybe a report would be a good idea.

Because if you set `diff-hl-flydiff-delay` to 2 before enabling this mode, you will see that the `a` stays near the edge of the screen for 2 seconds before flydiff does its update, and the indentation returns.

So unless I'm mistaken, right after typing `diff-hl` doesn't do anything in this scenario, and affects the buffer contents only after the delay has passed. That might mean that the bug is in `org-indent-mode` after all. Or in the display engine somewhere.

**Comment by dgutov** [comment]
> If diff-hl-flydiff-mode is disabled, no shift happens.

You can try creating an overlay like one `diff-hl-mode` adds, just without a modification hook (only a `before-string` an a `left-margin` instruction on there, maybe). That should make the problem appear too.

**Comment by peniblec** [comment]
OK, I'll report this with the following recipe:

- `emacs -Q`
- `C-x C-f repro.org RET`
- `M-x org-indent-mode RET`
- `* heading RET foo` (no RET yet)
- `M-:`

    ``` elisp
    (let ((ov (make-overlay (point-at-bol) (point-at-bol)))
          (val (propertize " " 'display '((left-fringe right-triangle)))))
      (overlay-put ov 'before-string val))
    ```

- `RET a`

Factlet Of The Day: typing characters with syntax `)` (closing delimiters watched by `blink-paren-post-self-insert-function`) causes the line-prefix to come back.

**Comment by dgutov** [comment]
> OK, I'll report this with the following recipe:

Looks good!

> Factlet Of The Day: typing characters with syntax ) (closing delimiters watched by blink-paren-post-self-insert-function) causes the line-prefix to come back

That's really weird. :-)

**Comment by dgutov** [comment]
Looks like it's fixed in Emacs master (to be released as 28.1 sometime next year, hopefully).

Thanks for all the investigative work!

**Comment by peniblec** [comment]
You're welcome! [I don't think we're completely out of the woods yet](https://debbugs.gnu.org/cgi/bugreport.cgi?bug=41584#14), but now it's clear that it's neither a diff-hl bug nor an org-indent one!
